Reuse SVG DOM elements for scatter traces

rreusser opened the pull request:
Second attempt. First was slightly botched and had more merge conflicts than changes.

rreusser commented:
Tests pass but not yet fit to merge. Need to add a couple additional sorts and solve a stale path data issue.

rreusser commented:
This is very nearly ready to go. The biggest challenge was visibility toggles since it tends to mess with SVG element ordering when traces are hidden and reshown. The solution is sorting, careful DOM maintenance, a bit of sorting, and some additional testing.

That said, I wouldn't be surprised if there are some corner cases that aren't tested and won't become apparent until this is really run through the wringer, so we'll have to exercise some caution if/when this is merged.

etpinard commented on src/plots/cartesian/index.js:
```diff
@@ -25,61 +25,74 @@ exports.attrRegex = constants.attrRegex;
+    } else {
+      // If traces are explicitly specified, then it's a partial replot and
+      // traces are not removed.
+        isFullReplot = false;
```
Hmm. Why do we need to handle the full-replot case in `Scatter.plot` instead of here in `Cartesian.plot` ?

rreusser replied:
I need to double check, but I ran into some weirdness where the scatter trace was handling ternary plots even though they're not actually Cartesian plots. I'll double check in just a couple minutes.

rreusser replied:
Disregard previous comment. Reason: was avoiding doing the keyed data join in `Cartesian.plot`. Since it's `Scatter.plot` that does it, it needs to know whether to remove un-updated traces.

etpinard commented on src/traces/scatter/attributes.js:
```diff
@@ -65,6 +65,10 @@ module.exports = {
     },
+    key: {
+        valType: 'data_array',
+        description: 'A list of keys for object constancy of data points during animation'
```
Let's remove `key` from the attributes for now.

Looks like it is only used internally at the moment (it is not getting coerced anywhere, right?).

I wouldn't want to pollute our attribute [schema](https://plot.ly/javascript/reference/) before implementing the functionality.

rreusser replied:
Works for me. Yeah, I didn't know quite where to draw the line removing features that will just need to be added right back in on the next step.

etpinard replied:
>  I didn't know quite where to draw the line removing features

No problem. :+1:

etpinard commented on src/traces/scatter/link_traces.js:
```diff
@@ -0,0 +1,43 @@
+
+        if(trace.visible === true && (
+            (Plots.traceIs(trace, 'cartesian') && trace.xaxis === xa._id && trace.yaxis === ya._id) ||
+            trace.type === 'scatterternary'
```
Looks like `linkTraces` is called from `Scatter.plot` only, so `cdsatter` is already guaranteed to pass the check :arrow_double_up: , right?

rreusser replied:
Yeah, I'd support thinking through this and paring it down to *only* checks that actually have any meaningful effect.

rreusser replied:
As long as cdscatter is guaranteed to be filtered properly to *only* scatter traces, then yes.

rreusser replied:
Removing the check. Mixed feelings. Should indeed be guaranteed correct based on usage, but also *must* be used correctly.

etpinard commented on src/traces/scatter/plot.js:
```diff
@@ -15,80 +15,158 @@ var Lib = require('../../lib');
-    var prevpath = '',
-        prevPolygons = [],
-        ownFillEl3, ownFillDir, tonext, nexttonext;
+    createFills(gd, scatterlayer);
```
Nice clean up. Thanks!

etpinard commented on src/traces/scatter/plot.js:
```diff
@@ -15,80 +15,158 @@ var Lib = require('../../lib');
+    // are shown and hidden. This is needed since we're not just wiping everything out
+    // and recreating on every update.
+    for(i = 0, uids = []; i < cdscatter.length; i++) {
+        uids[i] = cdscatter[i][0].trace.uid;
```
Looks good.

etpinard commented on src/traces/scatter/plot.js:
```diff
@@ -201,88 +287,131 @@ module.exports = function plot(gd, plotinfo, cdscatter) {
-            }
-        });
+    // XXX: Not okay. Just makes it work for now.
+    var i = 0;
```
Can you expand on :arrow_double_up:  ?

rreusser replied:
Ah, I forgot to return to that. The line below has `tracei.marker.maxdisplayed > 0 && j < i) {`, which I didn't understand (the `j < i` part). My PR changes how I was looping over these, so inter-trace logic like comparing the index didn't flow quite the same way. Maybe I can pass it the index of the trace though (the `i` in the removed line `cdscatter.forEach(function(d, i) {`) and it'll work the same way.

rreusser replied:
Let's see. I guess it would just be comparison between traces excluding comparison with self or duplicate comparisons with traces reversed… Ah, so it's counting up collisions, maybe?

etpinard replied:
I'm not familiar with this code. So I can't comment on what it _should_ do.

Maybe @alexcjohnson can help.

etpinard commented on test/image/mocks/ternary_simple.json:
```diff
@@ -16,8 +16,7 @@
                 2.12345
             ],
-            "type": "scatterternary",
-            "uid": "412fa8"
```
:beers:

etpinard commented on test/jasmine/tests/cartesian_test.js:
```diff
@@ -50,6 +49,104 @@ describe('zoom box element', function() {
+
+        afterEach(destroyGraphDiv);
+
+        it('reuses SVG fills', function(done) {
```
:+1:

etpinard commented on test/jasmine/tests/calcdata_test.js:
```diff
@@ -257,7 +257,7 @@ describe('calculated data and points', function() {
 
                 expect(gd.calcdata[0][0]).toEqual(jasmine.objectContaining({x: 6, y: 15}));
-                expect(gd.calcdata[0][1]).toEqual({x: false, y: false});
+                expect(gd.calcdata[0][1]).toEqual(jasmine.objectContaining({x: false, y: false}));
```
good

etpinard commented on src/plots/cartesian/index.js:
```diff
@@ -25,61 +25,74 @@ exports.attrRegex = constants.attrRegex;
+        // to update instead of fully removing and redrawing every time. The
+        // remaining plot traces should also be able to do this. Once implemented,
+        // we won't need this - which should sometimes be a big speedup.
+        if(subplotInfo.plot) {
```
Nice temporary workaround :+1: 

Could you add a comment on the line where class `trace` is added [here](https://github.com/plotly/plotly.js/pull/680/files/95f57a2751f8c8dae586ab42fcae76d6d2a15005#diff-ad4f76ccd6044ed16514297078e13b84R591) too?

rreusser replied:
👍🏼

etpinard commented:
@rreusser nice job! You made patches to some pretty gnarly code :beers: 

Everything looks ok at first glance. I'd say https://github.com/plotly/plotly.js/pull/680#discussion_r68766424 is the only blocking comment.

As you pointed out, our `restyle` has very poor test coverage at the moment. We should - at the very least - add tests for `restyle` calls that modify the scatter `mode` attribute.

rreusser commented:
@etpinard Q: do you think the right approach is to work with this a bit more, check things over, maybe add a few more tests, and then merge and release this? I think that makes the most sense so that if there's anything very wrong with this, we can catch it before there's also a lot of new stuff going on.

etpinard commented:
> and then merge and release this?

I would prefer merging it in `master` and let it sit there for about 1 week before releasing it.

rreusser commented:
Sounds good. I'm obv not up to speed on the release cycle yet, so I'll clean this up, start on the next stuff, and defer to you on that. 👍

rreusser commented:
@etpinard I've cleaned things up. That includes:

- did *not* move the `isFullReplot` logic inside scatter traces so that a data join is not necessary at the `Cartesian.plot` level.
- added note on scatter trace cleanup [here](https://github.com/plotly/plotly.js/pull/680/files?diff=split#diff-ad4f76ccd6044ed16514297078e13b84R587)
- removed redundant logic on filtering traces [here](https://github.com/plotly/plotly.js/pull/680/files?diff=split#diff-6de1ca97b8c54ce654b6176b5da3a98bR19) (so basically `linkTraces` must be used correctly, but it's only used in one place and the note is pretty clear about usage, so hopefully this is acceptable)
- had to pass some extra array references around for `selectMarkers` [here](https://github.com/plotly/plotly.js/pull/680/files?diff=split#diff-b15c528765724e5b795253c8314bb750R121), which is kind of ugly, but is probably better than an unnecessary refactor
- still some key logic in place [here](https://github.com/plotly/plotly.js/pull/680/files?diff=split#diff-b15c528765724e5b795253c8314bb750R303), but it just reduces to a no-op, so maybe this is acceptable.

rreusser commented:
PR https://github.com/plotly/plotly.js/pull/711 includes this PR and animation work, so as long as things can be tested adequately, there's probably not a great reason to merge this work until animations come together in their final form.
